File: SimulatorFrameworkBridgeLibTests/ServiceDispatchTests.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import <XCTest/XCTest.h>

#import <SimulatorFrameworkBridgeLib/ServiceDispatch.h>

@interface ServiceDispatchTests : XCTestCase
@end

@implementation ServiceDispatchTests

#pragma mark - Unknown service

- (void)testUnknownServiceReturnsFailure
{
  XCTAssertEqual(dispatchService(@"unknown", @"clear", @[]), 1);
}

- (void)testEmptyServiceReturnsFailure
{
  XCTAssertEqual(dispatchService(@"", @"clear", @[]), 1);
}

#pragma mark - Contacts routing

- (void)testDispatchContactsRoutes
{
  // Verifies routing reaches handleContactsAction (not "unknown service").
  // Return value depends on TCC state: 0 with authorization, 1 without.
  int result = dispatchService(@"contacts", @"clear", @[]);
  XCTAssertTrue(result == 0 || result == 1);
}

- (void)testDispatchContactsUnknownAction
{
  XCTAssertEqual(dispatchService(@"contacts", @"unknown", @[]), 1);
}

#pragma mark - Photos routing

- (void)testDispatchPhotosClearRoutes
{
  // Verifies routing reaches handlePhotoLibraryAction.
  // Result depends on photo library state and PLPhotoLibrary availability.
  int result = dispatchService(@"photos", @"clear", @[]);
  XCTAssertTrue(result == 0 || result == 1);
}

- (void)testDispatchPhotosUnknownAction
{
  XCTAssertEqual(dispatchService(@"photos", @"unknown", @[]), 1);
}

#pragma mark - Notifications routing

- (void)testDispatchNotificationsRoutes
{
  // BulletinBoard unavailable on macOS → returns 1
  XCTAssertEqual(dispatchService(@"notifications", @"approve", @[@"com.test"]), 1);
}

- (void)testDispatchNotificationsPassesBundleID
{
  // "check" with a bundleID — gateway fails but bundleID is passed through
  XCTAssertEqual(dispatchService(@"notifications", @"check", @[@"com.test"]), 1);
}

- (void)testDispatchNotificationsNoBundleID
{
  // "list" with no arguments — bundleID is nil
  XCTAssertEqual(dispatchService(@"notifications", @"list", @[]), 1);
}

#pragma mark - Proxy routing

- (void)testDispatchProxyMissingArgsRoutes
{
  // "set" with no args → insufficient arguments error
  XCTAssertEqual(dispatchService(@"proxy", @"set", @[]), 1);
}

- (void)testDispatchProxyUnknownAction
{
  XCTAssertEqual(dispatchService(@"proxy", @"unknown", @[]), 1);
}

#pragma mark - DNS routing

- (void)testDispatchDnsMissingArgsRoutes
{
  XCTAssertEqual(dispatchService(@"dns", @"set", @[]), 1);
}

- (void)testDispatchDnsUnknownAction
{
  XCTAssertEqual(dispatchService(@"dns", @"unknown", @[]), 1);
}

@end

File: SimulatorFrameworkBridgeLibTests/ServiceDispatchTests.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation
@_implementationOnly import SimulatorFrameworkBridgeLib
import XCTest

final class ServiceDispatchTests: XCTestCase {
  private var notificationsDirectory = ""

  override func setUp() {
    super.setUp()
    // What the delivered reader answers depends on the store it reads, so point it at an empty
    // directory of this suite's own: these assertions are about which service a verb reaches.
    notificationsDirectory = (NSTemporaryDirectory() as NSString)
      .appendingPathComponent(UUID().uuidString)
    FBDeliveredNotificationsSetDirectoryForTesting(notificationsDirectory)
    // The delivered verb reaches a real notification center on a simulator that has one, and
    // these assertions are about routing rather than about what it answers. Without a short
    // timeout, a center that does not come back spends the default 30s in a routing test.
    FBDeliveredNotificationsSetTimeoutForTesting(0.05)
  }

  override func tearDown() {
    FBDeliveredNotificationsSetDirectoryForTesting(nil)
    FBDeliveredNotificationsSetTimeoutForTesting(0)
    super.tearDown()
  }

  // MARK: - Unknown service

  func testUnknownServiceReturnsFailure() {
    XCTAssertEqual(dispatchService("unknown", "clear", []), 1)
  }

  // MARK: - Contacts routing

  func testDispatchContactsUnknownAction() {
    XCTAssertEqual(dispatchService("contacts", "unknown", []), 1)
  }

  // MARK: - Photos routing

  func testDispatchPhotosUnknownAction() {
    XCTAssertEqual(dispatchService("photos", "unknown", []), 1)
  }

  // MARK: - Notifications routing

  func testDispatchNotificationsRoutes() {
    XCTAssertEqual(dispatchService("notifications", "approve", ["com.test"]), 0)
  }

  func testDispatchNotificationsPassesBundleID() {
    XCTAssertEqual(dispatchService("notifications", "check", ["com.test"]), 0)
  }

  func testDispatchNotificationsNoBundleID() {
    // "list" with no arguments — bundleID is nil, so every section is reported
    XCTAssertEqual(dispatchService("notifications", "list", []), 0)
  }

  func testDispatchNotificationsDeliveredRoutesToTheReader() {
    // The settings service has no "delivered" action and would answer 1, so a 0 here is
    // the delivered reader answering.
    XCTAssertEqual(dispatchService("notifications", "delivered", ["com.test"]), 0)
  }

  // MARK: - Proxy routing

  func testDispatchProxyMissingArgsRoutes() {
    // "set" with no args → insufficient arguments error
    XCTAssertEqual(dispatchService("proxy", "set", []), 1)
  }

  func testDispatchProxyUnknownAction() {
    XCTAssertEqual(dispatchService("proxy", "unknown", []), 1)
  }

  // MARK: - DNS routing

  func testDispatchDnsMissingArgsRoutes() {
    XCTAssertEqual(dispatchService("dns", "set", []), 1)
  }

  func testDispatchDnsUnknownAction() {
    XCTAssertEqual(dispatchService("dns", "unknown", []), 1)
  }

  // MARK: - Health routing

  func testDispatchHealthRoutes() {
    // "list" for this bundle ID answers 1
    XCTAssertEqual(dispatchService("health", "list", ["com.test"]), 1)
  }

  func testDispatchHealthNoBundleID() {
    // "list" with no arguments — bundleID is nil, which answers 1
    XCTAssertEqual(dispatchService("health", "list", []), 1)
  }
}
